blog/2025-03-25-K3s-initialization.md

```diff
@@ -0,0 +1,119 @@
+
+```
+Starting k3s v1.32.2+k3s1
+Configuring sqlite3 database connection pooling: maxIdleConns=2, maxOpenConns=0, connMaxLifetime=0s
+Configuring database table schema and indexes, this may take a moment...
+Database tables and indexes are up to date
+Kine available at unix://kine.sock
+```
+
+Once the datastore is available k3s locks the bootstrap key. This step is useful for HA mode and this this key is just a placeholder so that other control-plane nodes do not start generating new CA certs. As we are not using HA mode in this example, this is not relevant.
+
+```
+Bootstrap key locked for initial create
+```
+
+K3s then generates all the TLS certificates required for the internal communications:
+
+```
+generated self-signed CA certificate
+certificate CN=system:admin,O=system:masters signed by CN=k3s-client-ca@1742309831
+certificate CN=system:k3s-supervisor,O=system:masters signed by CN=k3s-client-ca@1742309831
+certificate CN=system:kube-controller-manager signed by CN=k3s-client-ca@1742309831
+certificate CN=system:kube-scheduler signed by CN=k3s-client-ca@1742309831
+certificate CN=system:apiserver,O=system:masters signed by CN=k3s-client-ca@1742309831
+certificate CN=k3s-cloud-controller-manager signed by CN=k3s-client-ca@1742309831
+generated self-signed CA certificate CN=k3s-server-ca@1742309831
+certificate CN=kube-apiserver signed by CN=k3s-server-ca@1742309831
+generated self-signed CA certificate CN=k3s-request-header-ca@1742309831
+certificate CN=system:auth-proxy signed by CN=k3s-request-header-ca@1742309831
+generated self-signed CA certificate CN=etcd-server-ca@1742309831
+certificate CN=etcd-client signed by CN=etcd-server-ca@1742309831
+generated self-signed CA certificate CN=etcd-peer-ca@1742309831
+certificate CN=etcd-peer signed by CN=etcd-peer-ca@1742309831
+certificate CN=etcd-server signed by CN=etcd-server-ca@1742309831
+certificate CN=k3s,O=k3s signed by CN=k3s-server-ca@1742309831
+```
+
+And then saves the bootstrap data in the bootstrap key. Again, this is not relevant for this example:
+
+```
+Saving cluster bootstrap data to datastore
+```
+
+After that, K3s starts the different Kubernetes components. These components are all run within the k3s process as goroutines, which are lightweight, concurrent functions in Go, allowing for efficient resource usage. This is another design decision taken to reduce boot time and reduce resource consumption.
+
+```
+Running kube-apiserver
+Running kube-scheduler
+Running kube-controller-manager
+```
+
+Just for completeness, the charts are extracted in the directory `/var/lib/rancher/k3s/server/manifests/` when all the different Kubernetes components are running and K3s initialization is ready. The Helm Controller, which is a K3s embedded controller and its explanation is out of the scope for this post, will install any chart located in that directory. This is how components like CoreDNS or Traefik eventually get installed
```

Comment:
Technically the helm controller does not interact with this directory. The deploy controller watches this directory and applies all the manifests. If the manifest contains a HelmChart or HelmChartConfig resource, the helm controller will pick that up once the resource has been created in the cluster. The helm controller does not at any point interact directly with files in this directory.

Reply:
You are 100% right, thanks!